Add watch-only mode

jarys opened the pull request:
closes #238

### Changes
Note: by being optional I mean replacing type `T` with type `Option<T>`.

- `UnifiedSpendAuthority` replaced with `WalletCapability`
    - See #238 for more details.
- `WalletCapability as ReadableWriteable<()>` version incremented to 2, while staying compatible with version 1 `UnifiedSpendCapability as ReadableWriteable<()>` 
- `LightClient::create_with_seedorkey_wallet` renamed to `LightClient::new_from_wallet_base`
    - and its argument `seed_or_key: Option<String>` replaced with `wallet_base: WalletBase`
```rust
pub enum WalletBase {
    FreshEntropy,
    SeedBytes([u8; 32]),
    MnemonicPhrase(String),
    Mnemonic(Mnemonic),
    Ufvk(String),
}
```
- `UnifiedSpendAuthority: Into<SomeKey>` replaced with `WalletCapability: TryInto<SomeKey>`
    - i.e. all these conversions are `WalletCapability` failable
- arguments `orchard_ivk` and `sapling_ivk` of `TrialDecryptions::trial_decrypt_batch` are optional now
- field `mnemonic` of `LightWallet` is now optional
- `LightWallet::serialized_version()` incremented to 27 to enable optional mnemonic serialization
- `LightWallet::send_to_address_inner` returns error in watch-only mode
- field `extsk` of `SpendableSaplingNote` and field `spend_key` of `SpendableOrchardNote` are now optional
- added `ExtendedPubKey` struct as complement of `ExtendedPrivKey`
- added `secp256k1::SecretKey: ReadableWriteable<()>`
- added `OrchardSpendingKey: ReadableWriteable<()>`
- added `SaplingExtendedSpendingKey: ReadableWriteable<()>`
- added `--viewing-key` argument for `zingo-cli` :sparkles:

jarys commented:
PR is ready for review @AloeareV

commits should be be squashed before merging

there are no test scenarios at the moment . I will add some , this or next week . I tested on my own and everything was working as expected :sparkles: 

cc: @zancas

zancas commented on zingolib/src/wallet/keys/unified.rs:
```diff
@@ -84,53 +116,82 @@ fn read_write_receiver_selections() {
+            | (desired_receivers.sapling & !self.sapling.can_view()
+                | (desired_receivers.orchard & !self.orchard.can_view()))
+        {
+            return Err("Wallet is no capable to produce desired receivers".to_string());
```
"Wallet is not capable of producing desired receivers."

jarys replied:
3803f14f23f0bbf7229ca7878ffbdd86fba887db

zancas commented on zingolib/src/wallet/keys/unified.rs:
```diff
@@ -5,33 +5,65 @@ use std::{
-    sapling_key: zcash_primitives::zip32::ExtendedSpendingKey,
-    transparent_parent_key: super::extended_transparent::ExtendedPrivKey,
-    transparent_child_keys: Vec<(usize, secp256k1::SecretKey)>,
+pub enum Capability<ViewingKeyType, SpendKeyType> {
```
@AloeareV proposes making this `#[non_exhaustive']` both the future proof against `SpendViaHWW(ViewingKeyType)`, and to future proof against allowing different types of viewing keys `Outgoing` `Incoming`.

jarys replied:
3a73b7f11c9338ac302895471672502f197eca96

zancas commented on zingolib/src/wallet/keys/unified.rs:
```diff
@@ -84,53 +116,82 @@ fn read_write_receiver_selections() {
+        if self.transparent.can_spend() {
+            Ok(&self.transparent_child_keys)
+        } else {
+            Err("Wallet is not capable to spend transparent funds.".to_string())
```
"The wallet is not capable of spending transparent funds."

jarys replied:
a84c39daf77360b612a7bd532058753248b77793

AloeareV reviewed (changes requested):
Just a couple notes, this looks great! Needs tests, but I'd be happy to include those in a subsequent PR in order to get this in before more merge conflicts crop up.

zancas commented on zingolib/src/lightclient.rs:
```diff
@@ -497,37 +501,26 @@ impl LightClient {
                 ));
             }
         }
+        Runtime::new().unwrap().block_on(async move {
```
Nice!

jarys replied:
👌

zancas reviewed (commented):
This looks awesome.   We definitely need to add some tests.

zancas reviewed (changes requested):
We need to add tests that show proper handling of new-and-old wallet serialization formats.

jarys commented:
Sure, making `enum Capability` `#[non_exhaustive]` is definitely appropriate. I didn't know about this attribute.

jarys commented:
I will focus on tests now.

zancas commented:
One other ask I have is to make a type, for the Error variant of Results returned from public `fn`s.

jarys commented:
commits squashed . this is [no-op](https://github.com/zingolabs/zingolib/compare/7eab4a7..07bc566)

jarys commented:
conflicts resolved . this PR is ready to be merged

jarys commented:
I do force push one more time to rebase instead of merge.

jarys commented:
I found a mistake when I was writing tests. I was trying to load Sapling `ExtendedFullViewingKey`, but only `DiversifiableFullViewingKey` is included in UFVK.

jarys commented:
So I replaced Sapling Extended FVK with Diversifiable FVK (6cd0f4946e812a9a8cf455c9181c0e3ce9bb5831)

I removed `TransactionMetadatSet::adjust_spendable_status` (49cd06f4a20a8f8c2de9f3affa03052a742ebaad). This method requires `Fvk: PartialEq`, which is not implemented for Diversifiable FVK. But this method is outdated anyway. It was used in version 8 and below [[source](https://github.com/adityapk00/zecwallet-light-cli/blob/fd5d11f0f28e0f1628ea8fdad0cce16d50e6bb98/lib/src/lightwallet.rs#L383-L393)]

And I fixed `scenario_mempool_clearing` test (b71f4c42bbcd7907820792bbd1ae67007781e747). This test started failing, because it was using Debug print to compare curve points represented by non-normalized projective coordinates 🤷 yikwim

jarys commented:
Tests are finished. Blocked by
- #256

AloeareV commented:
@zancas @juanky201271 Note that this PR makes a change to the zingolib public interface that zingo-mobile uses, renames `LightClient::create_with_seedorkey_wallet` to `LightClient::new_from_wallet_base`, and it now takes a `WalletBase` instead of a `String` as an argument.
